colossalai/checkpoint_io/utils.py

```diff
@@ -77,55 +80,67 @@ def is_safetensor_checkpoint(checkpoint_file_path: str) -> bool:
 # ======================================
 # Helper functions for saving shard file
 # ======================================
-def shard_checkpoint(state_dict: torch.Tensor, max_shard_size: int = 1024, weights_name: str = WEIGHTS_NAME):
+def shard_checkpoint(state_dict: torch.Tensor, max_shard_size: int = 1024) -> Iterator[Tuple[OrderedDict, int]]:
  
     """
     Splits a model state dictionary in sub-checkpoints so that the final size of each sub-checkpoint does not exceed a
     given size.
     """
-    sharded_state_dicts = []
     current_block = {}
     current_block_size = 0
-    total_size = 0
 
     for key, weight in state_dict.items():
+        ret_block = None
+        ret_block_size = 0
         if type(weight) != DTensor:
             weight_size = calculate_tensor_size(weight)
 
             # If this weight is going to tip up over the maximal size, we split.
             if current_block_size + weight_size > max_shard_size:
-                sharded_state_dicts.append(current_block)
+                ret_block = current_block
+                ret_block_size = current_block_size
                 current_block = {}
                 current_block_size = 0
-
             current_block[key] = weight
             current_block_size += weight_size
-            total_size += weight_size
+            
+        if ret_block != None:
+            yield ret_block, ret_block_size
 
-    # Add the last block
-    sharded_state_dicts.append(current_block)
+    yield current_block, current_block_size
+    
 
+def build_index(state_dict_shard: Iterator[Tuple[OrderedDict, int]], shards_total_num: int, use_safetensors: bool, variant: str):
     # If we only have one shard, we return it
-    if len(sharded_state_dicts) == 1:
-        return {weights_name: sharded_state_dicts[0]}, None
+    weights_name = SAFE_WEIGHTS_NAME if use_safetensors else WEIGHTS_NAME
+    weights_name = add_variant(weights_name, variant)
+
+    save_index_file = SAFE_WEIGHTS_INDEX_NAME if use_safetensors else WEIGHTS_INDEX_NAME
+    save_index_file = add_variant(save_index_file, variant)
+
+    if shards_total_num == 1:
+        return {weights_name: next(state_dict_shard)[0]}, None
     
-    # Otherwise, let's build the index
     weight_map = {}
     shards = {}
-
-    for idx, shard in enumerate(sharded_state_dicts):
-        shard_file = weights_name.replace(".bin", f"-{idx+1:05d}-of-{len(sharded_state_dicts):05d}.bin")
+    total_size = 0
+    # shard_pair like (shard, shard_size)
+    for idx, shard_pair in enumerate(state_dict_shard):
+        shard_file = weights_name.replace(".bin", f"-{idx+1:05d}-of-{shards_total_num:05d}.bin")
         shard_file = shard_file.replace(
-            ".safetensors", f"-{idx + 1:05d}-of-{len(sharded_state_dicts):05d}.safetensors"
+            ".safetensors", f"-{idx + 1:05d}-of-{shards_total_num:05d}.safetensors"
```

Comment:
@FrankLeeeee Can the file be like `pytorch_model-00001.bin`? It wastes much time and memory to collect checkpoint twice for gemini.

Reply:
I think it's ok since `pytorch_model.bin.index.json` record the file name. What we have to do is to ensure the consistency.

Reply:
it is compatible with hugging face,   then we can load origin hf model.

Reply:
I tested hf `from_pretrained()`. This format is also supported.

![image](https://user-images.githubusercontent.com/23111350/233255106-a01707ba-9b84-41e3-a597-a4099bc29575.png)

![image](https://user-images.githubusercontent.com/23111350/233255133-5d9f9cef-c521-419e-a920-466e554184d9.png)

Reply:
in hf, when usesafetensors==True ,the checkpoint file likes "model.epoch-00001-of-00005.safetensors", while usesafetensors==False, checkpoint file likes "pytorch_model.epoch-00001-of-00005.bin"

Reply:
In both cases, the file names are read from index file, right?

Reply:
So does file name format really matter?

Reply:
yes， both read from index file. it matters, i proposed a fix to get this consistency.

Reply:
I mean file name format does not affect loading. Running the generator of shards twice to get the number of shards is a very luxury solution in MP/Zero scenario.

Reply:
You can just skip getting the number of shards and just name the checkpoint file like `pytorch_model-00001.bin` or `model-00001.safetensors`.